blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1033 @@
+
+        for (addr, _) in pending_hm {
+            out.append(
+                self.pending_for(&addr)
+                    .await
+                    .ok_or_else(|| Error::InvalidFromAddr)?
+                    .as_mut(),
+            )
+        }
+
+        let cur_ts = self.cur_tipset.read().await.clone();
+
+        Ok((out, cur_ts))
+    }
+
+    /// Return a Vector of signed messages for a given from address. This vector will be sorted by
+    /// each messsage's sequence. If no corresponding messages found, return None result type
+    async fn pending_for(&self, a: &Address) -> Option<Vec<SignedMessage>> {
+        let pending = self.pending.read().await;
+        let mset = pending.get(a)?;
+        if mset.msgs.is_empty() {
+            return None;
+        }
+        let mut msg_vec = Vec::new();
+        for (_, item) in mset.msgs.iter() {
+            msg_vec.push(item.clone());
+        }
+        msg_vec.sort_by_key(|value| value.message().sequence());
+        Some(msg_vec)
+    }
+
+    /// Return Vector of signed messages given a block header for self
+    pub async fn messages_for_blocks(
+        &mut self,
+        blks: &[BlockHeader],
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let mut msg_vec: Vec<SignedMessage> = Vec::new();
+
+        for block in blks {
+            let (umsg, mut smsgs) = self.api.read().await.messages_for_block(&block)?;
+
+            msg_vec.append(smsgs.as_mut());
+            for msg in umsg {
+                let mut bls_sig_cache = self.bls_sig_cache.write().await;
+                let smsg = recover_sig(&mut bls_sig_cache, msg).await?;
+                msg_vec.push(smsg)
+            }
+        }
+        Ok(msg_vec)
+    }
+
+    /// Return gas price estimate this has been translated from lotus, a more smart implementation will
+    /// most likely need to be implemented
+    pub fn estimate_gas_price(&self, nblocksincl: u64) -> Result<BigInt, Error> {
+        // TODO possibly come up with a smarter way to estimate the gas price
+        let min_gas_price = 0;
+        match nblocksincl {
+            0 => Ok(BigInt::from(min_gas_price + 2)),
+            1 => Ok(BigInt::from(min_gas_price + 1)),
+            _ => Ok(BigInt::from(min_gas_price)),
+        }
+    }
+
+    /// Load local messages into pending. As of  right now messages are not deleted from self's
+    /// local_message field, possibly implement this in the future?
+    pub async fn load_local(&mut self) -> Result<(), Error> {
+        let mut msg_vec = Vec::new();
+        while let Some(msg) = self.local_msgs.pop() {
+            if let Err(err) = self.add(&msg).await {
+                if err == Error::SequenceTooLow {
+                    warn!("error adding message: {:?}", err);
+                } else {
+                    msg_vec.push(msg);
+                }
+            } else {
+                msg_vec.push(msg);
+            }
+        }
+
+        self.local_msgs = msg_vec;
```

Comment:
You're removing values and reallocating a new vector without any reason (continuation of previous), if this remains the same, you should at very least set the capacity of the `msg_vec` so the vec doesn't keep getting reallocated as it hits capacity